src/tests/performance/ping_pong.rs

```diff
@@ -0,0 +1,200 @@
+        .await
+        .unwrap();
+    let node_addr = node.addr();
+
+    for synth_count in synth_counts {
+        // setup metrics recorder
+        let test_metrics = TestMetrics::default();
+        // clear metrics and register metrics
+        metrics::register_histogram!(METRIC_LATENCY);
+
+        let mut synth_handles = Vec::with_capacity(synth_count);
+        let test_start = tokio::time::Instant::now();
+        for _ in 0..synth_count {
+            synth_handles.push(tokio::spawn(simulate_peer(node_addr)));
+        }
+
+        // wait for peers to complete
+        for handle in synth_handles {
+            let _ = handle.await;
+        }
+
+        let time_taken_secs = test_start.elapsed().as_secs_f64();
+
+        let snapshot = test_metrics.take_snapshot();
+        if let Some(latencies) = snapshot.construct_histogram(METRIC_LATENCY) {
+            if latencies.entries() >= 1 {
+                // add stats to table display
+                table.add_row(RequestStats::new(
+                    synth_count as u16,
+                    PINGS,
+                    latencies,
+                    time_taken_secs,
+                ));
+            }
+        }
+    }
+
+    node.stop().unwrap();
+
+    // Display results table
+    println!("\r\n{}", table);
+}
+
+async fn simulate_peer(node_addr: SocketAddr) {
+    let config = TestConfig::default();
+    let mut synth_node = SyntheticNode::new(&config).await;
+
+    synth_node.connect(node_addr).await.unwrap();
+    let mut seq;
+
+    for _ in 0..PINGS {
+        // Generate unique sequence for each ping
+        seq = thread_rng().next_u32();
+
+        let payload = Payload::TmPing(TmPing {
+            r#type: PingType::PtPing as i32,
+            seq: Some(seq),
+            ping_time: None,
+            net_time: None,
+        });
+
+        // Send Ping
+        synth_node.unicast(node_addr, payload).unwrap();
+
+        let now = Instant::now();
+        let mut matched = false;
+
+        // There is a need to read messages in a loop as we can read message that is not ping reply.
+        while !matched {
+            match synth_node
+                .recv_message_timeout(Duration::from_secs(10))
+                .await
+            {
+                Ok(m) => {
+                    if matches!(
+                        &m.payload,
+                        Payload::TmPing(TmPing {
+                        r#type: r_type,
+                        seq: Some(s),
+                        ..
+                        }) if *s == seq && *r_type == PingType::PtPong as i32
+                    ) {
+                        metrics::histogram!(METRIC_LATENCY, duration_as_ms(now.elapsed()));
+                        matched = true;
+                    }
+                } // We can panic here as not receiving any message means that connection is dead.
+                Err(e) => panic!("Error receiving message - timeout?: {:?}", e),
+            }
```

Comment:
Shouldn't we actually `break` here instead? I see that's how it is implemented in the same Zcash test.
If we don't receive the message response, it may also mean that the rippled is running 100% and can't handle all requests on time.

Otherwise, checking `matched` in a `while` loop doesn't make much sense since it will never be set to `false` because we will always `panic` before

Reply:
And we are expecting one message every time we enter the loop, so not sure if we even need a `while` loop here.

Reply:
The sense of checking matched in loop is that we can receive messages which may be not ping reply. So then matches! macro return us false and matched remain false and new loop iteration will begin and new message will be received.

And you're right - we can panic in case of any other error than timeout and timeout could be for us only the way to "skip" this ping - when we break we can send new ping with new seq and check if it will be handled.

Reply:
> The sense of checking matched in the loop is that we can receive messages which may be not ping reply

Yeah, I realized that once I read the implementation for the recv_message_timeout function below 😆

> when we break we can send new ping with new seq and check if it will be handled.

yes, let's try like that, I'm wondering if it affects the results

Reply:
Changes done. It hasn't affected the results. In fact, CPU load is not a case - no CPU thread is fully loaded during test. Apparently, connections are refused earlier - I'll investigate this after creating pure connection test.